presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/MetadataQueryOptimizer.java

```diff
@@ -0,0 +1,192 @@
+        }
+
+        Map<Symbol, ColumnHandle> columns = columnBuilder.build();
+        Map<Symbol, Type> types = typesBuilder.build();
+
+        // Materialize the list of partitions and replace the TableScan node
+        // with a Values node
+        TableLayout layout = null;
+        if (!tableScan.getLayout().isPresent()) {
+            List<TableLayoutResult> layouts = metadata.getLayouts(session, tableScan.getTable(), Constraint.alwaysTrue(), Optional.empty());
+            if (layouts.size() == 1) {
+                layout = Iterables.getOnlyElement(layouts).getLayout();
+            }
+        }
+        else {
+            layout = metadata.getLayout(session, tableScan.getLayout().get());
+        }
+        if (layout == null || !layout.getDiscretePredicates().isPresent()) {
+            return Optional.empty();
+        }
+        DiscretePredicates predicates = layout.getDiscretePredicates().get();
+
+        // the optimization is only valid if the aggregation node only relies on partition keys
+        if (!predicates.getColumns().containsAll(columns.values())) {
+            return Optional.empty();
+        }
+
+        ImmutableList.Builder<List<Expression>> rowsBuilder = ImmutableList.builder();
+        for (TupleDomain<ColumnHandle> domain : predicates.getPredicates()) {
+            if (!domain.isNone()) {
+                Map<ColumnHandle, NullableValue> entries = TupleDomain.extractFixedValues(domain).get();
+
+                ImmutableList.Builder<Expression> rowBuilder = ImmutableList.builder();
+                // for each input column, add a literal expression using the entry value
+                for (Symbol input : inputs) {
+                    ColumnHandle column = columns.get(input);
+                    Type type = types.get(input);
+                    NullableValue value = entries.get(column);
+                    if (value == null) {
+                        // partition key does not have a single value, so bail out to be safe
+                        return Optional.empty();
+                    }
+                    else {
+                        rowBuilder.add(LiteralInterpreter.toExpression(value.getValue(), type));
+                    }
+                }
+                rowsBuilder.add(rowBuilder.build());
+            }
+        }
+
+        // replace the tablescan node with a values node
+        ValuesNode valuesNode = new ValuesNode(idAllocator.getNextId(), inputs, rowsBuilder.build());
+
+        PlanNode replacedNode = searchFrom(aggregation, lookup)
+                .where(TableScanNode.class::isInstance)
+                .replaceAll(valuesNode);
+        return Optional.of(replacedNode);
+    }
+
+    private static Optional<TableScanNode> findTableScan(PlanNode source, Lookup lookup)
+    {
+        // allow any chain of linear transformations
+        if (source instanceof MarkDistinctNode ||
+                source instanceof LimitNode ||
```

Comment:
I'm on the fence about whether limit should be allowed.  I can't think of a case where we would actually give wrong results because whenever we have limit with order by, we get a TopNNode, which is the only case I can think of as being a problem.  But still there's nothing that prevents a limit node from being on top of a sort, e.g. during a rewrite.